### ISSUE TITLE
pulsar-client-go reports message too large before compressing it.

### ISSUE THREAD

**Issue by flowchartsman** [apache/pulsar-client-go#447]
I'm getting "message too large" messages on some large JSON documents that, when compressed, are nowhere near the limit.

in `producer_partiton.go`:
```go
	// if msg is too large
	if len(payload) > int(p.cnx.GetMaxMessageSize()) {
		p.publishSemaphore.Release()
		request.callback(nil, request.msg, errMessageTooLarge)
		p.log.WithError(errMessageTooLarge).
			WithField("size", len(payload)).
			WithField("properties", msg.Properties).
			Error()
		p.metrics.PublishErrorsMsgTooLarge.Inc()
		return
	}
```

Correct me if I'm wrong, but this appears to check the payload size before any compression.

Whereas in `ProducerImpl.java`:
```java
int compressedSize = compressedPayload.readableBytes();
            if (compressedSize > ClientCnx.getMaxMessageSize() && !this.conf.isChunkingEnabled()) {
                compressedPayload.release();
```

**Comment by merlimat** [comment]
That's right, the Java and C++ clients are more permissive in that the message go through if it's below the limit after compression. We should also do the same thing in Go. 

Keep in mind, though, that it won't change the "guaranteed" max size, since there is guarantee that the compression will actually reduce the size.

**Comment by flowchartsman** [comment]
> Keep in mind, though, that it won't change the "guaranteed" max size, since there is guarantee that the compression will actually reduce the size.

I am not sure what this means. Of course there is never a guarantee with compression...

**Comment by merlimat** [comment]
Yes, if the max limit is 5MB, it will only guaranteed to accept up to 5MB. Above that, it's just "best effort" to get that to fit in with compression.

**Comment by flowchartsman** [comment]
Okay, so wait, I’m confused. I guess it might help to discuss the intended purpose of maxmessagesize.  If it’s just a knob to adjust cluster performance and storage allocation and such, that’s one thing, but if it is intended as a to guarantee that, for example, a consumer can keep a fixed buffer for handing the payload, then this would seem to be correct behavior, am I right? Are the Java and C++ clients actually *incorrect* in allowing message compression before payload size checking?

**Comment by freeznet** [comment]
@flowchartsman as my understanding, Java and C++ are doing the right way. `MaxMessageSize` can be considered as the buffer size between broker and client. But it also the buffer size between bookkeeper and pulsar. Pulsar and bookkeeper will not de-compress the message, and the decompression is a client side job. This is why we need to ensure that client side features are consistent across different languages.

**Comment by yuvalgut** [comment]
any news about this?

**Comment by cocktail828** [comment]
any updates on this issue? We meet the same problem recently.

**Comment by cocktail828** [comment]
hi everybody:
I just fix this issue this afternoon. Now in my fork, I move **compressionProvider** into _producer_partition.go_ from _batch_builder.go_, so I can compress payload before check the message size. 
[cocktail828/pulsar-client-go fixes apache#447](https://github.com/cocktail828/pulsar-client-go/commit/c0eeb48e750d6f4d82fcbfdc3daa3c1f96747111).

**Comment by flowchartsman** [comment]
This issue is now becoming a problem for us in golang pulsar functions again, due to occasional large JSON messages which, again, would likely compress very well, only they appear to be failing with `AddToBatchFailed` which is about as descriptive as saying "there was a failure".  Does _anyone_ have time to work on this one? Is there any interest in @cocktail828 's fix? Anyone?

**Comment by cocktail828** [comment]
> This issue is now becoming a problem for us in Go pulsar functions again, due to occasional large JSON messages which would likely compress very well, only they appear to be failing with `AddToBatchFailed` which is about as descriptive as saying "there was a failure". Does anyone have time to work on this one? Is there any interest in @cocktail828 's fix? I'm happy to clean it up and issue a PR, if necessary, so I guess I'll just do that if I don't hear anything else here first.

@flowchartsman Sorry for late reply, I have just issue a PR for this issue, please check that.
Currently golang SDK does not support 'chunk message' so the fix is much easier. Refer to c++ implement, it will check compressed-message size again when the raw message is too big.
